### DISCUSSION TITLE
Plan to cleanup the issue list

### DISCUSSION THREAD

**Discussion by jaap-karssenberg** [zim-desktop-wiki/zim-desktop-wiki#1610] [General]
Dear all,

I would like to discuss a proposal on how to cleanup the issue list. Below my intended plan of attack. Let me know any feedback in how to make it more effective.

Thanks,

Jaap

----

The current situation (September 2021) is that there is a backlog of ~ 2,5 year in issue triage. There are now ~ 600 issues open, of which ~ 400 uncategorized / not triaged, ~100 feature requests, ~25 bugs, ~25 usability, 50 other. Clearly this needs to be brought back to sanity in order for the issue list to be useful.

The desired situation would be to reduce the issue list to genuine bugs / issues with clear (reproduce-able) descriptions and work items / new features that are on the roadmap and will be worked on in the foreseeable future. The issue list should be leading in release planning.

Going forward questions and discussion of potential new features should be redirected to the discussions forum. I will update the project documentation & website to reflect this. The issue list remains open for anybody to file, but if should be limited to clear issues.

Next, the current backlog needs cleaning up - this is the ugly part, since it effectively means closing quite a few tickets even though the submitter may care deeply about it.

Candidates for closure are:
- Bug / issue reports that are vague, cannot be reproduced, lack information (after asking)
- Bug / issue that are not really and issue, e.g. Gtk warning only, opinion on aesthetics, etc.
- Feature requests without clear use case (redirect to discussions)
- Feature requests that do not fit my roadmap (*) or have an (active) contributor assigned
- Questions (redirect to dicussions)

Keep open:
- Issues to be fixed in the (near) future
- Issues that are clear, but solution is blocked - e.g. upstream issues - to avoid people filing the same issue again
- Feature requests that match my roadmap (*)
- Feature requests with a clear use case description that could be implemented as a 3rd party plugin

Process:
- For questions, try to answer and close immediately, add standard text to refer to discussion forum
- Issues / features on my roadmap (*) get a "`enhancement`" + "`roadmap idea`" label
- Feature requests that make sense as 3rd party plugin get a "`enhancement`" + "`plugin idea`" label
- Actual issues get either "`bug`", "`usability`", "`accessibility`" or "`documentation`" labels
- Issues that seem correct, but miss information get "`needs info`" label and may become candidate for closure in next cleaned round if the information is still lacking
- Issues that are reproduceable but are non-obvious get "`need investigation`" label
- Candidates for closure get label "`cleanup candidate`", than close if no new insight after ~3 month, add standard text with information about the process and refer to discussion forum

(*) the roadmap list is at my discretion, basically things I consider picking up myself, see 1.0.0 wishlist for examples [here](https://github.com/zim-desktop-wiki/zim-desktop-wiki/wiki/Planning#5-nice-to-haves-that-can-be-implemented-in-1x0-releases)

Actions:
- Add new issues for items on my worklist for next release(s), make plan more visible and allow more discussion upfront
- Write standard text for cleanup candidate issue / feature request & questions
- Update https://github.com/zim-desktop-wiki/zim-desktop-wiki/wiki/Issues and document above "rules"

**Comment by Simplenuity** [comment]
Very good idea and already good and detailed plan!
What made me wonder was putting these two types of feature requests on the issue list. I'd suggest to only put feature requests there that are actively worked on. Any other feature requests I'd put on the discussion list or, after and as the result of a discussion, put on the roadmap or some kind of waiting list. When something from the roadmap/waiting list gets picked up to actively being worked on it would then be moved over to the issue list.
Question regarding labels: Is it necessary/ would it be helpful to have labels for different Zim versions?

**Reply by sojusnik** [reply]
Support this idea! In `Discussions` we can have a dedicated section for feature requests, where people can discuss and give their vote for each request. Then top features could be picked and transferred to `Issues`, when the reaming features in `Issues` are resolved.

Here's an [example](https://github.com/iSoron/uhabits/discussions/categories/feature-requests) how this can look like.

**Comment by jaap-karssenberg** [comment]
Good point, for new feature requests start with discussion before moving to
issue list. Just want to keep the existing ones because they already
contain some information. Closing them now that information is lost unless
we keep track some way; so rather keep them open. Still it should be a
short list :)

Why would you want to label versions?  Future versions are tracked via
milestones. If bugs are reported against an old version we test against the
latest, no support for older versions. Could make sense to have a label for
issues that only occur in the development branch, since these should get
priority before releasing.

**Reply by Simplenuity** [reply]
"If bugs are reported against an old version we test against the
latest, no support for older versions." 
That answered the question. Labeling isn't necessary/helpful then.

**Comment by introt** [comment]
Nice plan! Since you're planning to keep issue submitting open to everybody, I'd like to suggest taking a look at issue templates as they're used by the [Xournal++ folk](https://github.com/xournalpp/xournalpp/issues/new/choose), pictured below:

![image](https://user-images.githubusercontent.com/26069787/134306855-67196552-e370-4ad6-ba84-f1fe97ce7d81.png)

When opening a new issue, you're first presented with templates to choose from. Templates may contain fields to fill or boxes to tick, eg. operating system and application version or "I have read [the issue guidelines]" as well as a few lines about looking for relevant issues and maybe opening a Discussion instead.

Regardless of whether you want the Templates to automatically apply certain labels or categorize issues manually, having (separate) templates for eg. bugs/crashes, usability/accessibility and documentation issues allows prompting the submitter for information relevant to solving the issue.

**Reply by introt** [reply]
By "the submitter" I especially mean the good-meaning non-technical folk, who after coming across a bug point their browsers at the repo, maybe create a GitHub account just for this one issue, click "New Issue" (without reading the README etc) and paste some lines from the GUI's exception handler with no further context - having a template in place gives them a chance to answer a few basic questions about their system and what they were doing then-and-there, before they log off and get on with their busy life (which may or may not include ever checking their email for follow-up :).

**Reply by introt** [reply]
Also, a separate template + tag for development branch bugs would IMO make a ton of sense

**Reply by silmaril42** [reply]
If this is used, I'd suggest adding a `needs triage` label to every new issue automatically - to be replaced by appropriate labels after Jaap (or some other competent person) had a look.

**Comment by Nick222** [comment]
I review all my issues and some of them need to convert in Discussion (I write it in its).

**Comment by jaap-karssenberg** [comment]
@Nick222 thanks for going over the list. Will not convert all of them them
immediately but intent to go over the list in batches. Will take your
comments into account when I get to those issues,

**Reply by introt** [reply]
I've labeled those with "candidate for removal" to not flood the discussions all at once

**Reply by introt** [reply]
Update: converting issues to discussions no longer puts them at the top of the Discussions list - this means we could start converting stale issues en masse to make the list more manageable.

**Comment by jaap-karssenberg** [comment]
Also implemented issue templates to structure new issues. Let's see how that works.

**Comment by jaap-karssenberg** [comment]
@introt a big Thank You for all the cleaning up you have done so far! Looks like the list is back in control now :)

**Reply by introt** [reply]
You're most welcome (also, welcome back)! Just trying to free up your time for my own issues >:3 There's still lots to do, but hopefully things will be a bit more manageable going forward.
